### DISCUSSION TITLE
Cryptography Next Generation APIs and PlatformNotSupportedException

### DISCUSSION THREAD

**Discussion by adamjstone** [dotnet/runtime#38197] [General] [answered]
# Question re. CNG APIs in .NET Standard

Documentation for numerous APIs in `System.Security.Cryptography.Cng` indicate that `PlatformNotSupportedException` is possible but do not list the formally supported platforms. Which platforms do not support use of the CNG APIs?

Thanks!

**Comment by vcsjones** [comment, marked as answer]
> Which platforms do not support use of the CNG APIs

CNG is Windows-specific. The CNG types (`RSACng`, etc) will only work on Windows, and throw `PlatformNotSupportedException` on non-Windows platforms.

If you want cross-platform cryptographic primitives, the factory methods on the primitives should be used. For example:

```c#
using (var rsa = System.Security.Cryptography.RSA.Create())
{
   //Use `rsa` here
}
```

and `Create` will give you an `RSA` implementation that will work for the current platform it is running on. The same approach works for other cryptographic primitives like `ECDsa`, `Aes`, etc.

**Reply by adamjstone** [reply]
Thank you.
